### ISSUE TITLE
`box.cfg` doesn't try to re-apply `listen` or `replication` options if the option value is the same

### ISSUE THREAD

**Issue by locker** [tarantool/tarantool#8551]
Reproduced with Tarantool 3.0.0-entrypoint-162-g22d7636b521f. Tarantool 2.10 and 2.11 are affected as well.

`box.cfg.listen` and `box.cfg.replication` aren't re-applied if the configuration value is the same.

```
tarantool> box.cfg{listen = 3301}
2023-04-11 15:14:41.011 [929508] main/103/interactive I> tx_binary: bound to 0.0.0.0:3301
2023-04-11 15:14:41.012 [929508] main/103/interactive/box.load_cfg I> set 'listen' configuration option to 3301
---
...

tarantool> box.cfg{listen = 3301}
---
...
```

Workaround: pass the same uri in a different way, for example, as a string:

```
tarantool> box.cfg{listen = 3301}
---
...

tarantool> box.cfg{listen = '3301'}
2023-04-11 15:17:18.168 [929508] main/103/interactive I> tx_binary: stopped
2023-04-11 15:17:18.168 [929508] main/103/interactive I> tx_binary: bound to 0.0.0.0:3301
2023-04-11 15:17:18.169 [929508] main/103/interactive/box.load_cfg I> set 'listen' configuration option to "3301"
---
...
```

This is confusing and sometimes may complicate operation. For example, if there's a uri parameter that points to a file (e.g. an SSL certificate), then after updating the file one can't just call `box.cfg` with the same uri to reload the file (https://github.com/tarantool/tarantool-ee/issues/432). Another example: to restart replication one has to reset the `replication` table first by calling `box.cfg{replication = {}}`.

We should apply all options passed to `box.cfg` unconditionally, even if the value is unchanged. It should be up to the underlying callback to decide whether it should do anything or not. For example, the replication configuration callback already doesn't restart replication for appliers that are up and running, see commit https://github.com/tarantool/tarantool/commit/5994892c2f2c34823356efe5d8560e31e9ad7c42, so this extra check in `load_cfg.lua` is pointless.

**Comment by Mons** [comment]
It seems that we need some kind of box.ctl.replication_restart and box.ctl.iproto_restart to restart the iproto and replication with current settings

**Comment by R-omk** [comment]
https://github.com/tarantool/tarantool/commit/5994892c2f2c34823356efe5d8560e31e9ad7c42

**Comment by locker** [comment]
> That's by design.
> It was not so in 1.6, and reapplying the same config invoked restart of replication. Then it was fixed.

After commit https://github.com/tarantool/tarantool/commit/5994892c2f2c34823356efe5d8560e31e9ad7c42 (pointed to by @R-omk) active replication won't be restarted on `box.cfg{replication = box.cfg.replication}`. It won't be restarted even if the uri order is changed. The old hack that compares the old and the new `box.cfg` values and skips the config update if they are the same looks pointless after that fix.

OTOH because of that fix one can't reload SSL certificates by calling `box.cfg{replication = box.cfg.replication}` even if this issue is fixed so to fix https://github.com/tarantool/tarantool-ee/issues/432 we would need something like `box.ctl.restart_replication`. Alternatively, we could make the replication configuration callback carefully check if SSL certificates were updated and restart replication only in this case.

**Comment by Totktonada** [comment]
I agree with @Mons's position in that it is quite natural to expect that unchanged option has no effect (by default at least). I agree with @locker in that sometimes it is necessary to ask for reconfiguration of some subsystem/parameter. And I agree that it is convenient to do in the same place, where you write a config, not using some separate call/signal.

I have an idea. How about a new box.cfg option that lists option that should be re-applied disregarding whether they're changed or not?

```lua
box.cfg({
    listen = <...>,
    reapply = {'listen'},
})
```

The name is arbitrary. Maybe `force` fits good here as well.

**Comment by locker** [comment]
> I have an idea. How about a new box.cfg option that lists options that should be re-applied disregarding whether they're changed or not?

Why not just let the configuration option callback decide whether the value has changed or not? The replication callback already handles unchanged URIs and doesn't restart replication of all URIs when, for example, a new URI added to the list. We could update the `listen` callback to do the same although I'm not sure this is necessary. As for other dynamic options callbacks, calling them unconditionally seem to be safe - they just have no effect if nothing's changed.

Comparing option values in Lua (`load_cfg.lua`) looks wrong because there we do this without the context. For example, rearranging URIs in `box.cfg.replication` should be a no-op while updating the SSL certificate file should trigger reconfiguration. Neither of these cases can be detected at the `load_cfg.lua` level.

**Comment by Mons** [comment]
> OTOH because of that fix one can't reload SSL certificates by calling `box.cfg{replication = box.cfg.replication}` even if this issue is fixed so to fix [tarantool/tarantool-ee#432](https://github.com/tarantool/tarantool-ee/issues/432) we would need something like `box.ctl.restart_replication`. Alternatively, we could make the replication configuration callback carefully check if SSL certificates were updated and restart replication only in this case.

Ok, got it. That's ok.
Moreover: It would be great to make host resolution, if peers are listed as hostnames.

`box.ctl.restart_replication` may be done separately as for another maintenance scenario.

> The issue title does look confusing though - like we're going to reconfigure replication anyway, not just call the cfg callback.

You are right, I thought it just rolled back old behavior with replication restart.

**Comment by unera** [comment]
I don't like adding option reapply. This interface is hard and unnecessary for users. I think we can write additional tests if we doubt in something.

**Comment by Mons** [comment]
> I have an idea. How about a new box.cfg option that lists options that should be re-applied disregarding whether they're changed or not?
> 
> ```lua
> box.cfg({
>     listen = <...>,
>     reapply = {'listen'},
> })
> ```
> 

That's not handy. There are either applying "new" configuration from configuration source (without knowledge whether changed it or not) or it may be maintenance operation done by hand without reconfiguration (and in that case touching the box.cfg would be not handy, just like now with `local r = box.cfg.replication; box.cfg{ replication={} } box.cfg{ replication = r}`

**Comment by locker** [comment]
> > The issue title does look confusing though - like we're going to reconfigure replication anyway, not just call the cfg callback.
> 
> You are right, I thought it just rolled back old behavior with replication restart.

I updated the issue title and description. Hopefully, it's less confusing now.

**Comment by Totktonada** [comment]
> As for other dynamic options callbacks, calling them unconditionally seem to be safe - they just have no effect if nothing's changed.

My doubts are mostly about this part. I suspect that it may be error-prone (because a developer may not expect it). But we can list options that perform its own compare explicitly, that's implementation detail. Okay so.

**Comment by locker** [comment]
Some vshard and cartridge tests seem to depend on the existing behavior and assume that appliers won't even try to connect to the target instances (to check their uuids) when `box.cfg.replication` is reapplied (`box.cfg{replication = box.cfg.replication}`). We need to investigate if this is just a test issue, in which case we can just fix the tests, or a show-stopper, in which case we need to either come up with another way of updating SSL certificates or fix the `box.cfg.replication` callback to avoid reconenct/auth if uri is the same. I filed two issues to investigate this: https://github.com/tarantool/vshard/issues/411 https://github.com/tarantool/cartridge/issues/2073.

**Comment by R-omk** [comment]
Related https://github.com/tarantool/tarantool/issues/7805

**Comment by locker** [comment]
Re-applying `box.cfg.replication` on `box.cfg{replication = box.cfg.replication}` turns out to be too risky, see https://github.com/tarantool/tarantool/issues/8551#issuecomment-1505503984. We decided that we shouldn't do anything if the URI set is the same. Still, we need to move the check from Lua to C so that we can handle the case when URIs are the same but SSL certificate files they point to needs to be reloaded. This will be a pure refactoring without any user-visible changes. Closing this as "Won't Fix" for now.
